### ISSUE TITLE
python-docx destroys {PAGE} and {NUMPAGES} fields in header and footer during manipulation

### ISSUE THREAD

**Issue by jrp13149** [python-openxml/python-docx#686]
This is a great tool for automatically generating documents. I have read that it doesn't handle automated page numbering, because that is something that is rendered at runtime, but I've been trying to figure out a way to work around that.

I've used the following to create an empty template:

`from docx import Document`
`import re`
`document = Document()`
`paragraph = document.add_paragraph('Lorem ipsum dolor sit amet.')`
`section = document.sections[0]`
`footer=section.footer`
`paragraph = footer.paragraphs[0]`
`paragraph.text = "Hopefully this is a footer"`
`out='./outputs/test.docx'`
`document.save(out)`

and the test.docx file it creates is attached. I then mod the file in word, by going in and manually adding a 'Page X of Y' field in the footer, test_modded.docx, also attached.

[test.docx](https://github.com/python-openxml/python-docx/files/3301919/test.docx)
[test_modded.docx](https://github.com/python-openxml/python-docx/files/3301921/test_modded.docx)

My thinking is that while python-docs doesn't know about the page numbering fields, I could perhaps use the modded document as a template and do a search and replace on the text component, to check whether the page numbering fields can survive being passed through python-docx.

`moddeddocument = Document('./outputs/test_modded.docx')`
`section = moddeddocument.sections[0]`
`footer=section.footer`
`paragraph = footer.paragraphs[0]`
`paragraph.text=paragraph.text.replace("Hopefully", "Thankfully")`
`moddeddocument.save('./outputs/test_modded_out.docx')`

[test_modded_out.docx](https://github.com/python-openxml/python-docx/files/3301945/test_modded_out.docx)

Unfortunately, these fields don't survive, and get flatted to ASCII.

It seems like the act of evaluating parargaph.text is what destroys the control codes I want to preserve.
Any thoughts on how I might preserve them under the kind of operation I describe above?

Thanks,

Jeff.

**Comment by scanny** [comment]
Your problem is using the `.text` "utility" property which basically operates on the "nuke and replace" principle; (which is quite often a great strategy, just not in this case).

So you'll have to work at the run level, which brings the added complication that Word likes to break these up in a way that I'm sure makes sense to it but generally appears arbitrary to any other observer. So a good place to start is by printing the XML to the console and then working out how to edit it in a way that doesn't disrupt the fields.

```python
print(paragraph._p.xml)
```

If you can "curate" the template, then manually editing the XML to place all the text-to-be-changed in say, the first run, then `paragraph.runs[0].text = "New text"` could work.

If you can paste the paragraph XML and say more about your situation, like what inputs it needs to be able to handle I can probably give you more to go on.

**Comment by jrp13149** [comment]
Thanks!

I’m helping my wife compile school reports for her class. She has a large spreadsheet, one sheet per curriculum area, each with information about all the children.

I’ve written a Jupyter notebook to read the sheets into a dataframe, and have a curated docx template which I’ve inserted tokens into ($fullname, $mathscore, etc)

When run, the notebook iterates through each child, starting with the template and replacing the tokens with information, saving the docx to a new file, named after the child.

All works well except if a pile of reports falls on the floor it’s impossible to work out which page belongs to which child, so I want a footer template of the form “End of year report for $childname - Page X of Y”

I can find and replace $childname but so far it destroys the page numbers. I will give the xml approach a go in the morning!

Regards,

Jeff.

Sent from my iPhone

On 18 Jun 2019, at 18:12, Steve Canny <notifications@github.com<mailto:notifications@github.com>> wrote:


Your problem is using the .text "utility" property which basically operates on the "nuke and replace" principle; (which is quite often a great strategy, just not in this case).

So you'll have to work at the run level, which brings the added complication that Word likes to break these up in a way that I'm sure makes sense to it but generally appears arbitrary to any other observer. So a good place to start is by printing the XML to the console and then working out how to edit it in a way that doesn't disrupt the fields.

print(paragraph._p.xml)

If you can "curate" the template, then manually editing the XML to place all the text-to-be-changed in say, the first run, then paragraph.runs[0].text = "New text" could work.

If you can paste the paragraph XML and say more about your situation, like what inputs it needs to be able to handle I can probably give you more to go on.

**Comment by scanny** [comment]
Yeah, so that definitely sounds doable. Worst case you could put the footer content into a table and then you could do `paragraph.text` in one cell without affecting the right-hand cell at all.

**Comment by jrp13149** [comment]
So, I've had some success in seeing what I'm trying to modify, but don't see how to write the modified string back to the document. My code looks like this:

    `section = document.sections[0]`
    `footer=section.footer`
    `para = footer.paragraphs[0]`
    `new_para=re.sub('fullname', fullname, para._p.xml)`    
    `para._p.xml = new_para`

para._p.xml returns a lot of xml, mostly meaningless to me, but I can see the target token in there that I want to change, and re.sub can successfully change the token, creating a replacement string,

But the last command, trying to assign the new bundle of xml back to the paragraph, fails with:

`---------------------------------------------------------------------------`
`AttributeError                            Traceback (most recent call last)`
`<ipython-input-24-628fea196428> in <module>`
`     84     #print(para.text)`
`     85     new_para=re.sub('fullname', fullname, para._p.xml)`
`---> 86     para._p.xml = new_para`

`AttributeError: can't set attribute`

This approach works with the .text utility. Is there an alternative that could work here?

Thanks,

Jeff.

**Comment by scanny** [comment]
I need to see the paragraph XML to help you. Just paste in what comes out of this:

```python
print(paragraph._p.xml)
```

**Comment by tac11tac** [comment]
Granted, figuring out how to do it is the fun part.

However, if your wife just needs it done, then the standard word mailmerge
tools would do what you want.

Todd

Todd A Carter
734.323.7963
tac11tac at gmail dot com

**Comment by jrp13149** [comment]
> I need to see the paragraph XML to help you. Just paste in what comes out of this:
> 
> ```python
> print(paragraph._p.xml)
> ```

Ok, here is the snippet I'm running:
```python
# try for the footer
    section = document.sections[0]
    footer=section.footer
    para = footer.paragraphs[0]
    print(para._p.xml)
    new_para=re.sub('fullname', fullname, para._p.xml)    
    print(new_para)
#    para._p.xml = new_para
```
Here is the output of the first print statement - the target token 'fullname' is pretty easy to find. Below that is the output of the second statement, with the successful substitution. The issue is that I need to assign the modified string to a new string, but I don't see how to put the new string back into 'para'. 

<w:p xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main" xmlns:w14="http://schemas.microsoft.com/office/word/2010/wordml" xmlns:wpc="http://schemas.microsoft.com/office/word/2010/wordprocessingCanvas" xmlns:cx="http://schemas.microsoft.com/office/drawing/2014/chartex" xmlns:cx1="http://schemas.microsoft.com/office/drawing/2015/9/8/chartex" xmlns:cx2="http://schemas.microsoft.com/office/drawing/2015/10/21/chartex" xmlns:cx3="http://schemas.microsoft.com/office/drawing/2016/5/9/chartex" xmlns:cx4="http://schemas.microsoft.com/office/drawing/2016/5/10/chartex" xmlns:cx5="http://schemas.microsoft.com/office/drawing/2016/5/11/chartex" xmlns:cx6="http://schemas.microsoft.com/office/drawing/2016/5/12/chartex" xmlns:cx7="http://schemas.microsoft.com/office/drawing/2016/5/13/chartex" xmlns:cx8="http://schemas.microsoft.com/office/drawing/2016/5/14/chartex" xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" xmlns:aink="http://schemas.microsoft.com/office/drawing/2016/ink" xmlns:am3d="http://schemas.microsoft.com/office/drawing/2017/model3d" xmlns:o="urn:schemas-microsoft-com:office:office" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" xmlns:m="http://schemas.openxmlformats.org/officeDocument/2006/math" xmlns:v="urn:schemas-microsoft-com:vml" xmlns:wp14="http://schemas.microsoft.com/office/word/2010/wordprocessingDrawing" xmlns:wp="http://schemas.openxmlformats.org/drawingml/2006/wordprocessingDrawing" xmlns:w10="urn:schemas-microsoft-com:office:word" xmlns:w15="http://schemas.microsoft.com/office/word/2012/wordml" xmlns:w16cid="http://schemas.microsoft.com/office/word/2016/wordml/cid" xmlns:w16se="http://schemas.microsoft.com/office/word/2015/wordml/symex" xmlns:wpg="http://schemas.microsoft.com/office/word/2010/wordprocessingGroup" xmlns:wpi="http://schemas.microsoft.com/office/word/2010/wordprocessingInk" xmlns:wne="http://schemas.microsoft.com/office/word/2006/wordml" xmlns:wps="http://schemas.microsoft.com/office/word/2010/wordprocessingShape" w14:paraId="75970544" w14:textId="4D81D48D" w:rsidR="00472A0D" w:rsidRPr="005B2AD4" w:rsidRDefault="00A31394">
  <w:pPr>
    <w:pStyle w:val="Footer"/>
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
  </w:pPr>
  <w:r>
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:t xml:space="preserve">Crawley Ridge Infant School Annual Report – </w:t>
  </w:r>
  <w:proofErr w:type="spellStart"/>
  <w:r w:rsidR="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:t>fullname</w:t>
  </w:r>
  <w:proofErr w:type="spellEnd"/>
  <w:r>
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:tab/>
  </w:r>
  <w:r w:rsidR="005B2AD4" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:t>P</w:t>
  </w:r>
  <w:r w:rsidR="000F4A76" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:t xml:space="preserve">age </w:t>
  </w:r>
  <w:r w:rsidR="000F4A76" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:fldChar w:fldCharType="begin"/>
  </w:r>
  <w:r w:rsidR="000F4A76" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:instrText xml:space="preserve"> PAGE \* MERGEFORMAT </w:instrText>
  </w:r>
  <w:r w:rsidR="000F4A76" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:fldChar w:fldCharType="separate"/>
  </w:r>
  <w:r w:rsidR="000F4A76" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
      <w:noProof/>
    </w:rPr>
    <w:t>1</w:t>
  </w:r>
  <w:r w:rsidR="000F4A76" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:fldChar w:fldCharType="end"/>
  </w:r>
  <w:r w:rsidR="008B5BF6">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:t xml:space="preserve"> </w:t>
  </w:r>
  <w:r w:rsidR="000F4A76" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:t xml:space="preserve">of </w:t>
  </w:r>
  <w:r w:rsidR="0034235D" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:fldChar w:fldCharType="begin"/>
  </w:r>
  <w:r w:rsidR="0034235D" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:instrText xml:space="preserve"> NUMPAGES \*MERGEFORMAT </w:instrText>
  </w:r>
  <w:r w:rsidR="0034235D" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:fldChar w:fldCharType="separate"/>
  </w:r>
  <w:r w:rsidR="000F4A76" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
      <w:noProof/>
    </w:rPr>
    <w:t>1</w:t>
  </w:r>
  <w:r w:rsidR="0034235D" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
      <w:noProof/>
    </w:rPr>
    <w:fldChar w:fldCharType="end"/>
  </w:r>
</w:p>

Modified string:

<w:p xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main" xmlns:w14="http://schemas.microsoft.com/office/word/2010/wordml" xmlns:wpc="http://schemas.microsoft.com/office/word/2010/wordprocessingCanvas" xmlns:cx="http://schemas.microsoft.com/office/drawing/2014/chartex" xmlns:cx1="http://schemas.microsoft.com/office/drawing/2015/9/8/chartex" xmlns:cx2="http://schemas.microsoft.com/office/drawing/2015/10/21/chartex" xmlns:cx3="http://schemas.microsoft.com/office/drawing/2016/5/9/chartex" xmlns:cx4="http://schemas.microsoft.com/office/drawing/2016/5/10/chartex" xmlns:cx5="http://schemas.microsoft.com/office/drawing/2016/5/11/chartex" xmlns:cx6="http://schemas.microsoft.com/office/drawing/2016/5/12/chartex" xmlns:cx7="http://schemas.microsoft.com/office/drawing/2016/5/13/chartex" xmlns:cx8="http://schemas.microsoft.com/office/drawing/2016/5/14/chartex" xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" xmlns:aink="http://schemas.microsoft.com/office/drawing/2016/ink" xmlns:am3d="http://schemas.microsoft.com/office/drawing/2017/model3d" xmlns:o="urn:schemas-microsoft-com:office:office" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" xmlns:m="http://schemas.openxmlformats.org/officeDocument/2006/math" xmlns:v="urn:schemas-microsoft-com:vml" xmlns:wp14="http://schemas.microsoft.com/office/word/2010/wordprocessingDrawing" xmlns:wp="http://schemas.openxmlformats.org/drawingml/2006/wordprocessingDrawing" xmlns:w10="urn:schemas-microsoft-com:office:word" xmlns:w15="http://schemas.microsoft.com/office/word/2012/wordml" xmlns:w16cid="http://schemas.microsoft.com/office/word/2016/wordml/cid" xmlns:w16se="http://schemas.microsoft.com/office/word/2015/wordml/symex" xmlns:wpg="http://schemas.microsoft.com/office/word/2010/wordprocessingGroup" xmlns:wpi="http://schemas.microsoft.com/office/word/2010/wordprocessingInk" xmlns:wne="http://schemas.microsoft.com/office/word/2006/wordml" xmlns:wps="http://schemas.microsoft.com/office/word/2010/wordprocessingShape" w14:paraId="75970544" w14:textId="4D81D48D" w:rsidR="00472A0D" w:rsidRPr="005B2AD4" w:rsidRDefault="00A31394">
  <w:pPr>
    <w:pStyle w:val="Footer"/>
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
  </w:pPr>
  <w:r>
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:t xml:space="preserve">Crawley Ridge Infant School Annual Report – </w:t>
  </w:r>
  <w:proofErr w:type="spellStart"/>
  <w:r w:rsidR="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:t>Jeff Parke</w:t>
  </w:r>
  <w:proofErr w:type="spellEnd"/>
  <w:r>
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:tab/>
  </w:r>
  <w:r w:rsidR="005B2AD4" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:t>P</w:t>
  </w:r>
  <w:r w:rsidR="000F4A76" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:t xml:space="preserve">age </w:t>
  </w:r>
  <w:r w:rsidR="000F4A76" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:fldChar w:fldCharType="begin"/>
  </w:r>
  <w:r w:rsidR="000F4A76" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:instrText xml:space="preserve"> PAGE \* MERGEFORMAT </w:instrText>
  </w:r>
  <w:r w:rsidR="000F4A76" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:fldChar w:fldCharType="separate"/>
  </w:r>
  <w:r w:rsidR="000F4A76" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
      <w:noProof/>
    </w:rPr>
    <w:t>1</w:t>
  </w:r>
  <w:r w:rsidR="000F4A76" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:fldChar w:fldCharType="end"/>
  </w:r>
  <w:r w:rsidR="008B5BF6">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:t xml:space="preserve"> </w:t>
  </w:r>
  <w:r w:rsidR="000F4A76" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:t xml:space="preserve">of </w:t>
  </w:r>
  <w:r w:rsidR="0034235D" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:fldChar w:fldCharType="begin"/>
  </w:r>
  <w:r w:rsidR="0034235D" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:instrText xml:space="preserve"> NUMPAGES \*MERGEFORMAT </w:instrText>
  </w:r>
  <w:r w:rsidR="0034235D" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
    </w:rPr>
    <w:fldChar w:fldCharType="separate"/>
  </w:r>
  <w:r w:rsidR="000F4A76" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
      <w:noProof/>
    </w:rPr>
    <w:t>1</w:t>
  </w:r>
  <w:r w:rsidR="0034235D" w:rsidRPr="005B2AD4">
    <w:rPr>
      <w:rFonts w:asciiTheme="majorHAnsi" w:hAnsiTheme="majorHAnsi" w:cstheme="majorHAnsi"/>
      <w:noProof/>
    </w:rPr>
    <w:fldChar w:fldCharType="end"/>
  </w:r>
</w:p>

**Comment by jrp13149** [comment]
> Granted, figuring out how to do it is the fun part. However, if your wife just needs it done, then the standard word mailmerge tools would do what you want.

Fought mailmerge last year - very flaky. This is so much better.

**Comment by scanny** [comment]
Wow. That's huge :)  Looks like Word's spell-checker has had its way with it.

I've reformatted it above by enclosing it in triple backticks with the "xml" syntax highlighting keyword. Now that it's nicely formatted, note that the enclosing paragraph element `<w:p>` has mostly run (`<w:r>`) elements for children. This might help you see what's happening inside, it decomposes quite cleanly.

So there are a couple quick solutions I can see, something like this:

```python
runs = footer.paragraphs[0].runs
runs[1].text = "New Name"
runs[4].text = "age 3"
# ---etc.---
```

This combined with a trick to remove an unwanted run may get you going:

```python
unwanted_run = runs[6]  # ---or whatever index you want to get rid of---
r = unwanted_run._r
r.getparent().remove(r)
```
Basically you edit convenient runs by assigning new text and delete any extras you find yourself with. As long as you don't modify the page-number-field runs they'll be just fine.

Note this may change the indices of following runs, so be alert for that.

Something like this can help you get the run index right:

```python
for idx, run in enumerate(paragraph.runs):
    print("run[%d].text == %s" % (idx, run.text))
```

Maybe see how far you can get with that and we can go from there. The next step up in sophistication is to manually curate the "template" file so the runs exactly contain each "field" and are in a very predictable position. But no need for that if this will get it done.

**Comment by jrp13149** [comment]
Brilliant. This worked:

`    section = document.sections[0]`
`    footer=section.footer`
`    para = footer.paragraphs[0]`
`    runs = para.runs`
`    runs[1].text = fullname`

I'll add in a loop to make sure it's robust to any future changes, but it's great to know it works.

Thanks!

**Comment by scanny** [comment]
Glad you got it working :)

**Comment by yungztrunks** [comment]
6 years later this thread helped me solve a very similar issue! I had a merge field in the footer which upon replacement corrupted the page count on the bottom right of the footer (from `1/16`, `2/16`, `3/16` etc.. to `/116`, `/216`, `/316` etc..). Thanks a lot!
